content/en/docs/Reference/declarative-config.md

```diff
@@ -0,0 +1,341 @@
+- Promoting an existing bundle to a new channel
+- Changing the default channel of a package
+- Adding, updating, or removing upgrade edges
+
+### Composability
+
+Declarative Configs are stored in an arbitrary directory hierarchy, which enables index composition. If I have two separate DC directories, `indexA` and `indexB`, I can make a new combined index by making a new directory `indexC` and copying `indexA` and `indexB` into it.
+
+This composability enables decentralized indexes. The format permits operator authors to maintain operator-specific indexes and index maintainers to trivially build an index composed of individual operator indexes.
+
+One of the major benefits is that those who are most familiar with an operator, its dependencies, and its upgrade compatibility (i.e. the operator authors) are able to maintain their own operator-specific index and have direct control over its contents. DC moves the task of building and maintaining indexes more towards operator authors, thus giving composite index maintainers more time to build value around their compositions.
+
+Another benefit is that the format enables composite index maintainers to build their index without the knowledge or coordination of the maintainers of the sub-indexes. Indexes like this can be composed by combining multiple other indexes or by extracting only necessary subsets of one index, or a combination of both of these.
+
+### Extensibility
+
+Another motivation is to enable more extensibility around indexes. DC is a low-level representation of an index. While it can be maintained directly in its low-level form, we expect many index maintainers to build interesting extensions on top that can be used by their own custom tooling to make all sorts of mutations. For example, one could imagine a tool that translates a high-level API like (mode=semver) down to the low-level DC format for upgrade edges. Or perhaps an index maintainer needs to customize all of the bundle metadata by adding a new property to bundles that meet a certain criteria.
+
+The OLM developer community will be making use of this extensibility to build more official tooling on top of the low-level APIs, but the major benefit is that index maintainers have this capability as well.
+
+## Specification
+
+### Structure
+
+Declarative config can be stored and loaded from directory-based filesystems.
+
+`opm` loads declarative config by walking the root directory and recursing into subdirectories. It attempts to load every file it finds as declarative config and fails if any errors occur.
+
+Non-DC files can be ignored using `.indexignore` files, which behave identically to `.gitignore` files. That is, they have the same rules for [patterns](https://git-scm.com/docs/gitignore#_pattern_format) and precedence.
+
+> #### Example `.gitignore` file
+> ```gitignore
+> # Ignore everything except non-object .json and .yaml files
+> **/*
+> !*.json
+> !*.yaml
+> **/objects/*.json
+> **/objects/*.yaml
+> ```
```

Comment:
We probably want to have `.indexignore` on this explicitly, for anyone not familiar with `.gitignore` who'd expect opm to ignore it automatically.

Reply:
This particular set of patterns _would_ ignore the .indexignore file due to the `**/*` pattern, which ignores everything (the follow-on patterns walk that back a little so that non-object yaml files are _not_ ignored).

However, we updated `opm` [here](https://github.com/operator-framework/operator-registry/pull/733) to automatically include `.indexignore` in the set of ignored files.